pkg/i2gw/ingress2gateway.go

```diff
@@ -127,3 +132,64 @@ func GetSupportedProviders() []string {
+//   - Gateways may have the same NamespaceName even if they come from different
+//     ingresses, as they have a their GatewayClass' name as name. For this reason,
+//     if there are mutiple gateways named the same, their listeners are merged into
+//     a unique Gateway.
+func MergeGatewayResources(gatewayResources ...GatewayResources) (GatewayResources, field.ErrorList) {
+	mergedGatewayResources := GatewayResources{
+		Gateways:        make(map[types.NamespacedName]gatewayv1.Gateway),
+		GatewayClasses:  make(map[types.NamespacedName]gatewayv1.GatewayClass),
+		HTTPRoutes:      make(map[types.NamespacedName]gatewayv1.HTTPRoute),
+		TLSRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TLSRoute),
+		TCPRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TCPRoute),
+		UDPRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.UDPRoute),
+		ReferenceGrants: make(map[types.NamespacedName]gatewayv1alpha2.ReferenceGrant),
+	}
+	var errs field.ErrorList
+	mergedGatewayResources.Gateways, errs = mergeGateways(gatewayResources)
+	if len(errs) > 0 {
+		return GatewayResources{}, errs
+	}
+	for _, gr := range gatewayResources {
+		maps.Copy(mergedGatewayResources.GatewayClasses, gr.GatewayClasses)
+		maps.Copy(mergedGatewayResources.HTTPRoutes, gr.HTTPRoutes)
+		maps.Copy(mergedGatewayResources.TLSRoutes, gr.TLSRoutes)
+		maps.Copy(mergedGatewayResources.TCPRoutes, gr.TCPRoutes)
+		maps.Copy(mergedGatewayResources.UDPRoutes, gr.UDPRoutes)
+		maps.Copy(mergedGatewayResources.ReferenceGrants, gr.ReferenceGrants)
+	}
+	return mergedGatewayResources, errs
+}
+
+func mergeGateways(gatewaResources []GatewayResources) (map[types.NamespacedName]gatewayv1.Gateway, field.ErrorList) {
+	newGateways := map[types.NamespacedName]gatewayv1.Gateway{}
+	errs := field.ErrorList{}
+
+	for _, gr := range gatewaResources {
+		for _, g := range gr.Gateways {
+			nn := types.NamespacedName{Namespace: g.Namespace, Name: g.Name}
+			if existingGateway, ok := newGateways[nn]; ok {
+				g.Spec.Listeners = append(g.Spec.Listeners, existingGateway.Spec.Listeners...)
+				g.Spec.Addresses = append(g.Spec.Addresses, existingGateway.Spec.Addresses...)
+			}
+			newGateways[nn] = g
+			// 64 is the maximum number of listeners a Gateway can have
+			if len(g.Spec.Listeners) > 64 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("listeners")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 64 listeners"))
+			}
+			// 16 is the maximum number of addresses a Gateway can have
+			if len(g.Spec.Addresses) > 16 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("addresses")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 16 addresses"))
+			}
+		}
+	}
+
+	return newGateways, errs
+}
```

Comment:
Mind adding a comment that this behavior is likely to change following https://github.com/kubernetes-sigs/gateway-api/pull/1863/ ?

Reply:
added